### PR TITLE
8317422: [JVMCI] concurrency issue in MethodData creation

### PR DESCRIPTION
**Pull request by dougxc**
`Method::build_profiling_method_data` is the safe way to create `MethodData` for a `Method` - it uses an atomic to do the write.
This PR updates `c2v_getFailedSpeculationAddress` and `c2v_reprofile` to use `Method::build_profiling_method_data`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317422](https://bugs.openjdk.org/browse/JDK-8317422): [JVMCI] concurrency issue in MethodData creation (**Bug** - P2)

### Reviewers
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**) ⚠️ Review applies to [fda052d4](https://git.openjdk.org/jdk/pull/16026/files/fda052d4394d3a57f7fa16fca13c409bc14a78b0)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16026/head:pull/16026` \
`$ git checkout pull/16026`

Update a local copy of the PR: \
`$ git checkout pull/16026` \
`$ git pull https://git.openjdk.org/jdk.git pull/16026/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16026`

View PR using the GUI difftool: \
`$ git pr show -t 16026`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16026.diff">https://git.openjdk.org/jdk/pull/16026.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16026#issuecomment-1746430605)

### CODE REVIEW

**Comment by tkrodriguez** [comment]
It's possible my diagnosis of this is wrong since `set_method_data` tries to handle the memory ordering by doing a `release_store` but maybe that's not strong enough?  I see there is also `release_store_fence`.  Anyway, I think this cleanup is a good idea even if it's not the full fix.

I think we should eliminate the public `set_method_data` completely.  `set_method_data(null)` should be a new `clear_method_data()` and we'd need to fix ciReplay.cpp to also call `build_profiling_method_data`.

**Comment by dougxc** [comment]
> I think we should eliminate the public set_method_data completely. set_method_data(null) should be a new clear_method_data() and we'd need to fix ciReplay.cpp to also call build_profiling_method_data.

Good idea. I made these changes.

File: src/hotspot/share/oops/method.hpp [hunk 1]
```diff
@@ -308,7 +308,7 @@ class Method : public Metadata {
     return _method_data;
   }
 
-  void set_method_data(MethodData* data);
+  void clear_method_data();
```
Comment (tkrodriguez) [inline src/hotspot/share/oops/method.hpp:311]: This should be private now I think.

Comment (dougxc) [reply src/hotspot/share/oops/method.hpp]: Makes sense - done. I also inlined its definition into `method.hpp`.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by dougxc** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**